powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationBehavior.java

```diff
@@ -0,0 +1,469 @@
+            if (!tmpTokenOptional.isPresent()) {
+                operationId = tmpOperationId;
+                break;
+            } // ... else this token ID has a collision, reset it and try to find another one
+        }
+        if (operationId == null) {
+            logger.error("Unable to generate token");
+            // Rollback is not required, error occurs before writing to database
+            throw localizationProvider.buildExceptionForCode(ServiceError.UNABLE_TO_GENERATE_TOKEN);
+        }
+
+        // Get operation expiration date
+        final long expiration = templateEntity.getExpiration() * 1000L;
+        final Date timestampExpiration = new Date(currentTimestamp.getTime() + expiration);
+
+        // Build operation data
+        final StringSubstitutor sub = new StringSubstitutor(parameters);
+        final String operationData = sub.replace(templateEntity.getDataTemplate());
+
+        // Serialize parameters in the string
+        String parametersString;
+        try {
+            parametersString = objectMapper.writeValueAsString(parameters);
+        } catch (JsonProcessingException ex) { // Should not happen
+            parametersString = "{}";
+            logger.error("Unable to serialize JSON parameter payload for operation {}.", operationId, ex);
+        }
+
+        // Create a new operation
+        final OperationEntity operationEntity = new OperationEntity();
+        operationEntity.setId(operationId);
+        operationEntity.setUserId(userId);
+        operationEntity.setApplicationId(applicationId);
+        operationEntity.setTemplate(templateEntity);
+        operationEntity.setExternalId(externalId);
+        operationEntity.setOperationType(templateEntity.getOperationType());
+        operationEntity.setData(operationData);
+        operationEntity.setParameters(parametersString);
+        operationEntity.setStatus(OperationStatusDo.PENDING);
+        operationEntity.setSignatureType(templateEntity.getSignatureType());
+        operationEntity.setFailureCount(0L);
+        operationEntity.setMaxFailureCount(templateEntity.getMaxFailureCount());
+        operationEntity.setTimestampCreated(currentTimestamp);
+        operationEntity.setTimestampExpires(timestampExpiration);
+        operationEntity.setTimestampFinalized(null); // empty initially
+
+        final OperationEntity savedEntity = operationRepository.save(operationEntity);
+        return convertFromEntity(savedEntity);
+
+    }
+
+    public OperationUserActionResponse attemptApproveOperation(OperationApproveRequest request) throws GenericServiceException {
+        final Date currentTimestamp = new Date();
+
+        final String operationId = request.getOperationId();
+        final String userId = request.getUserId();
+        final Long applicationId = request.getApplicationId();
+        final String data = request.getData();
```

Comment:
We should either add validation to the requests or check for null value.

Reply:
We will fix this when removing the SOAP interface by using well-annotated request objects. The good part is that these services are never called publically, always from trusted systems.